File: app/javascript/components/features/Result.js
import React from 'react';
import json from '../../../../public/education_cost.json';
import WithoutTitleCard from '../components/WithoutTitleCard';
import Text from '../components/Text';
import ResultGraph from './ResultGraph';
import Link from '../components/Link';
import Button from '../components/Button';

const Result = ({ props }) => {
  //積立期間
  const duration = (18 - props.age) * 12

  //教育費算出
  //costArrayはグラフに渡す、３年ごとの教育費累計額を格納した配列
  //costDatasは年齢別に教育費を格納した配列
  const SchoolTypes = [
    {type: 'nurserySchool', age: 6, term: ['age1', 'age2', 'age3', 'age4', 'age5']},
    {type: 'kindergarten', age: 6, term: ['age3', 'age4', 'age5']},
    {type: 'primarySchool', age: 12, term: ['age6', 'age7', 'age8', 'age9', 'age10', 'age11']},
    {type: 'juniorHighSchool', age: 15, term: ['age12', 'age13', 'age14']},
    {type: 'highSchool', age: 18, term: ['age15', 'age16', 'age17']},
    {type: 'university', age: 22, term: ['age18', 'age19', 'age20', 'age21'],}
  ]

  const costDatas = [0];
  const costArray = [0];
  let total = 0;
  SchoolTypes.forEach((school) => {
    if (props[school.type] !== 'unselected'){
      if (props.age < school.age){
        const categoryKey = props[school.type];
        const publicOrPrivate = json[school.type][categoryKey];

        let ageRange;
        if (school.type === 'nurserySchool' && props['kindergarten'] !== 'unselected'){
          if (props.entryage === '1'){
            ageRange = ['age1', 'age2']
          } else if (props.entryage === '2'){
            ageRange = ['age2']
          } else {
            ageRange = []
          }
        } else if (school.type === 'kindergarten' && props['kindergarten'] === 'unselected'){
          ageRange = [];
        } else {
          ageRange = school.term
        }

        ageRange.forEach((age) => {
          let livingAloneCost = 0;
          if (school.type === 'university' && props.livingAloneFunds !== 0){
            age === 'age18'
            ? livingAloneCost = publicOrPrivate[age] + props.livingAloneFunds * 12 + json['livingAllowance']['initialize']
            : livingAloneCost = publicOrPrivate[age] + props.livingAloneFunds * 12;
          }
          const cost = publicOrPrivate[age] + livingAloneCost;
          if ((parseInt(age.slice(3), 10) + 1) % 3 === 0 && parseInt(age.slice(3), 10) !== 20){
            total += cost;
            costArray.push(total)
          } else if (parseInt(age.slice(3)) === 21){
            total += cost;
            costArray.push(total)
          } else {
            total += cost;
          }
          costDatas.push(cost);
        })
      } else {
        for(let i = 0; i < school.term.length; i++){
          if (school.type !== 'kindergarten'){
            costDatas.push(0);
            const eachAge = parseInt(school.term[i].slice(3), 10);
            if ((eachAge + 1) % 3 === 0 && eachAge !== 20){
              costArray.push(0);
            } else if (eachAge === 21) {
              costArray.push(0);
            }
          }
        }
      }
    } else {
      if (school.type === 'nurserySchool') {
        costArray.push(0);
        Array.from({ length: 2 }, (_, index) => {
          costDatas.push(0);
        })
      }
    }
  })

  //教育費の総額
  const amount = costDatas.reduce(function(a,b){
    return a + b;
  })
  const formattedAmount = amount.toLocaleString();


  //月々の積立金額
  const monthlyAmount = Math.round(amount / duration);
  const formattedMonthlyAmount = monthlyAmount.toLocaleString();

  return (
    <div>
      <WithoutTitleCard widthSize='1/2' bgColor='blue-100' borderColor='blue-500'>
        <div>
          <Text size='lg' color='amber-dark' style='mb-3'>{props.age}〜18才まで　{duration}ヶ月</Text>
          <Text size='3xl' color='amber-dark' style='mb-3'>総額： {formattedAmount}円</Text>
          <Text size='lg' color='amber-dark'>月額　約{formattedMonthlyAmount}円</Text>
        </div>
      </WithoutTitleCard>

      <div className='w-1/2 m-auto'>
        <ResultGraph monthlyAmount={monthlyAmount} age={props.age} costArray={costArray} />
      </div>

      <div className='mt-3'>
        <Text size='xs' color='amber-dark'>
          『令和3年度子供の学習費調査の結果について（文部科学省より・令和4年12月発行）』
        </Text>
        <Text size='xs' color='amber-dark'>
          『令和3年度「教育費負担の実態調査結果」（日本政策金融公庫より・令和3年12月発行）』
        </Text>
      </div>
      <div className='mt-10'>
        <p className='text-center text-amber-dark'>会員登録をして、教育費を貯めましょう。</p>
        <Link page='/users/new'>
          <Button pxSize='3' pySize='2' color='green-300' fontColor='white' roundType='full'>会員登録（無料）</Button>
        </Link>
      </div>
    </div>
  );
}

export default Result;

File: app/models/result.rb
class Result < ApplicationRecord
  belongs_to :child
  BLANK = -1
  MAX_AGE = 18
  MONTHS_PER_YEAR = 12
  YEARS_FOR_UNI = 4
  TEN_THOUSAND = 10_000
  HIGH_SCHOOL_ENTRY_COST = 350_000

  validates :age, presence: true, numericality: { greater_than_or_equal_to: 0 }
  validates :nursery_school, presence: true
  validates :from_age_for_nursery_school, numericality: { greater_than_or_equal_to: 0 }
  validates :kindergarten, presence: true
  validates :primary_school, presence: true
  validates :junior_high_school, presence: true
  validates :high_school, presence: true
  validates :university, presence: true
  validates :living_alone_funds, presence: true, numericality: { greater_than_or_equal_to: 0 }

  def self.params_format(result_params)
    age = result_params[:age]
    result_params[:age] = if age == '選択してください'
                            BLANK
                          else
                            age == '出産前' ? 0 : age.to_i
                          end
    result_params[:from_age_for_nursery_school] = 0 if result_params[:nursery_school] == 'unselected'
    living_alone_funds = result_params[:living_alone_funds]
    result_params[:living_alone_funds] = if living_alone_funds == '選択してください'
                                           BLANK
                                         else
                                           living_alone_funds == '仕送りの予定はない' ? 0 : living_alone_funds.delete(',').to_i
                                         end

    result_params
  end

  def self.json_cost_datas
    # JSONからデータ取得
    cost_datas = {}
    File.open('./public/education_cost.json') do |file|
      cost_datas = JSON.load(file)
    end
  end

  def self.school_types
    schoolTypes = %w[
      nursery_school
      kindergarten
      primary_school
      junior_high_school
      high_school
      university
    ]
    schoolTypes
  end

  def age_range
    ageRange = {
      nursery_school: %w[age1 age2],
      kindergarten: %w[age3 age4 age5],
      primary_school: %w[age6 age7 age8 age9 age10 age11],
      junior_high_school: %w[age12 age13 age14],
      high_school: %w[age15 age16 age17],
      university: %w[age18 age19 age20 age21]
    }
    if unselected?('kindergarten')
      array = []
      (from_age_for_nursery_school..5).each do |target_age|
        array.push("age#{target_age}")
      end
      ageRange[:nursery_school] = array
      ageRange[:kindergarten] = []
    end
    ageRange
  end

  def duration
    (MAX_AGE - age) * MONTHS_PER_YEAR
  end

  def total_cost
    cost_datas_hash.values.sum
  end

  def cost_per_month
    total_cost / duration
  end

  def cost_datas_hash
    json_datas = Result.json_cost_datas
    cost_datas = {}
    schoolTypes = [
      'nurserySchool',
      'kindergarten',
      'primarySchool',
      'juniorHighSchool',
      'highSchool',
      'university'
    ]
    
    ageRange = age_range

    schoolTypes.each do |schoolType|
      total = 0
      school_type = schoolType.underscore.to_sym
      unless unselected?(school_type)
        ageRange[school_type].each do |target_age|
          data = json_datas[schoolType][send(school_type)][target_age]
          total += data if target_age.gsub(/[^\d]/, "").to_i >= age
        end
      end
      cost_datas[school_type] = total
    end

    cost_datas[:living_alone_funds] =
      living_alone_funds.zero? ? 0 : json_datas['livingAllowance']['initialize'] + total_living_alone_funds
    cost_datas
  end

  def cost_datas_by_age
    json_datas = Result.json_cost_datas
    cost_datas = {}
    schoolTypes = %w[
      nurserySchool
      kindergarten
      primarySchool
      juniorHighSchool
      highSchool
      university
    ]
    ageRange = age_range

    schoolTypes.each do |schoolType|
      school_type = schoolType.underscore.to_sym
      unless unselected?(school_type)
        ageRange[school_type].each do |target_age|
          data = json_datas[schoolType][send(school_type)][target_age]
          target_age.gsub(/[^\d]/, "").to_i >= age ? cost_datas[target_age.to_sym] = data : cost_datas[target_age.to_sym] = 0
        end
      end
    end
    cost_datas[:living_alone_initialize] = living_alone_funds.zero? ? 0 : json_datas['livingAllowance']['initialize']
    cost_datas[:living_alone_funds] = total_living_alone_funds
    cost_datas
  end

  def each_stage_cost
    result_hash = {}

    # 高校入学費用
    result_hash[:high_school_cost] = HIGH_SCHOOL_ENTRY_COST

    # 大学入学費用
    university_entry_cost = { publicArts: 672_000, publicScience: 672_000, privateArts: 818_000, privateScience: 888_000 }
    unselected?('university') ? result_hash[:university_cost] = 0 : result_hash[:university_cost] = university_entry_cost[university.to_sym]
    result_hash
  end

  def total_living_alone_funds
    living_alone_funds * TEN_THOUSAND * MONTHS_PER_YEAR * YEARS_FOR_UNI
  end

  def unselected?(school_type)
    return send(school_type) == 'unselected'
  end

  def cost_until_age(target_age)
    datas = cost_datas_by_age.to_a
    total = 0
    datas.each do |data|
      if data[0].to_s.gsub(/[^\d]/, "").to_i <= target_age
        total += data[1] if data[0].to_s.gsub(/[^\d]/, "").to_i >= age
      end
    end
    total
  end

  def cost_array_for_graph
    array = [0]
    (1..18).each do |idx|
      if idx % 3 == 0
        array.push(cost_until_age(idx))
      end
    end
    array.push(cost_until_age(22))
  end
end
